File: examples/nodl_tutorials/lifecycle_action/cpp/fibonacci_action_server.cpp
// SPDX-FileCopyrightText: 2026 Open Source Robotics Foundation, Inc.
// SPDX-License-Identifier: Apache-2.0

#include <atomic>
#include <chrono>
#include <functional>
#include <memory>
#include <thread>

#include "example_interfaces/action/fibonacci.hpp"
#include "fibonacci_action_server_cpp_base.hpp"  // NOLINT(build/include_subdir)
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp"

using std::chrono_literals::operator""ms;

class FibonacciActionServer : public FibonacciActionServerCppBase
{
  using Fibonacci = example_interfaces::action::Fibonacci;
  using GoalHandle = rclcpp_action::ServerGoalHandle<Fibonacci>;
  using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;

protected:
  CallbackReturn on_activate(const rclcpp_lifecycle::State & previous_state) override
  {
    const auto result = FibonacciActionServerCppBase::on_activate(previous_state);
    active_.store(result == CallbackReturn::SUCCESS);
    return result;
  }

  CallbackReturn on_deactivate(const rclcpp_lifecycle::State & previous_state) override
  {
    active_.store(false);
    return FibonacciActionServerCppBase::on_deactivate(previous_state);
  }

  rclcpp_action::GoalResponse on_fibonacci_goal(
    const rclcpp_action::GoalUUID & uuid, std::shared_ptr<const Fibonacci::Goal> goal) override
  {
    (void)uuid;
    RCLCPP_INFO(get_logger(), "Received goal request with order %d", goal->order);
    if (!active_.load() || goal->order > 9000) {
      return rclcpp_action::GoalResponse::REJECT;
    }
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  rclcpp_action::CancelResponse on_fibonacci_cancel(std::shared_ptr<GoalHandle> goal_handle) override
  {
    (void)goal_handle;
    RCLCPP_INFO(get_logger(), "Received request to cancel goal");
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  void on_fibonacci_accepted(std::shared_ptr<GoalHandle> goal_handle) override
  {
    std::thread{std::bind(&FibonacciActionServer::execute, this, std::placeholders::_1), goal_handle}.detach();
  }

private:
  void execute(const std::shared_ptr<GoalHandle> goal_handle)
  {
    const auto goal = goal_handle->get_goal();
    auto feedback = std::make_shared<Fibonacci::Feedback>();
    feedback->sequence = {0, 1};
    auto result = std::make_shared<Fibonacci::Result>();

    for (int i = 1; i < goal->order && rclcpp::ok(); ++i) {
      if (goal_handle->is_canceling()) {
        result->sequence = feedback->sequence;
        goal_handle->canceled(result);
        return;
      }
      feedback->sequence.push_back(
        feedback->sequence[static_cast<std::size_t>(i)] + feedback->sequence[static_cast<std::size_t>(i - 1)]);
      goal_handle->publish_feedback(feedback);
      std::this_thread::sleep_for(100ms);
    }

    if (rclcpp::ok()) {
      result->sequence = feedback->sequence;
      goal_handle->succeed(result);
    }
  }

  std::atomic_bool active_{false};
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FibonacciActionServer>()->get_node_base_interface());
  rclcpp::shutdown();
  return 0;
}

File: examples/nodl_tutorials/lifecycle_action/python/fibonacci_action_server.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2026 Open Source Robotics Foundation, Inc.
# SPDX-License-Identifier: Apache-2.0
"""Lifecycle Fibonacci server implemented on the generated Python base."""

import time

import rclpy
from example_interfaces.action import Fibonacci
from nodl_tutorial_lifecycle_action.generated.fibonacci_action_server_py_base import (
    FibonacciActionServerPyBase,
)
from rclpy.action import CancelResponse, GoalResponse
from rclpy.executors import MultiThreadedExecutor
from rclpy.lifecycle import TransitionCallbackReturn


class FibonacciActionServer(FibonacciActionServerPyBase):
    def __init__(self):
        super().__init__()
        self._active = False

    def on_activate(self, state):
        result = super().on_activate(state)
        self._active = result == TransitionCallbackReturn.SUCCESS
        return result

    def on_deactivate(self, state):
        self._active = False
        return super().on_deactivate(state)

    def on_fibonacci_goal(self, goal_request):
        self.get_logger().info(f'Received goal request with order {goal_request.order}')
        if not self._active or goal_request.order > 9000:
            return GoalResponse.REJECT
        return GoalResponse.ACCEPT

    def on_fibonacci_cancel(self, goal_handle):
        del goal_handle
        self.get_logger().info('Received request to cancel goal')
        return CancelResponse.ACCEPT

    def execute_fibonacci(self, goal_handle):
        feedback = Fibonacci.Feedback(sequence=[0, 1])
        for index in range(1, goal_handle.request.order):
            if goal_handle.is_cancel_requested:
                goal_handle.canceled()
                return Fibonacci.Result(sequence=feedback.sequence)
            feedback.sequence.append(feedback.sequence[index] + feedback.sequence[index - 1])
            goal_handle.publish_feedback(feedback)
            time.sleep(0.1)

        goal_handle.succeed()
        return Fibonacci.Result(sequence=feedback.sequence)


def main(args=None):
    rclpy.init(args=args)
    node = FibonacciActionServer()
    executor = MultiThreadedExecutor(num_threads=2)
    try:
        rclpy.spin(node, executor=executor)
    finally:
        executor.shutdown()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
